qiskit/algorithms/evolvers/pvqd.py

```diff
@@ -0,0 +1,309 @@
+                "parameter shift rule to the ansatz."
+            ) from exc
+
+        # check whether all parameters are unique and we do not need to apply the chain rule
+        # (since it's not implemented yet)
+        all_parameters = []
+        for circuit_instruction in unrolled.data:
+            for param in circuit_instruction.operation.params:
+                if isinstance(param, ParameterExpression):
+                    if isinstance(param, Parameter):
+                        all_parameters.append(param)
+                    else:
+                        raise QiskitError(
+                            "Circuit is only allowed to have plain parameters, as "
+                            "the chain rule is not yet implemented."
+                        )
+
+        if len(all_parameters) != self.ansatz.num_parameters:
+            raise QiskitError(
+                "Circuit parameters must be unique, the product rule is not yet " "implemented."
+            )
+
+    def _get_observable_evaluator(self, observables):
+        if isinstance(observables, list):
+            observables = ListOp(observables)
+
+        expectation_value = StateFn(observables, is_measurement=True) @ StateFn(self.ansatz)
+        converted = self.expectation.convert(expectation_value)
+
+        ansatz_parameters = self.ansatz.parameters
+
+        def evaluate_observables(theta: np.ndarray) -> Union[float, List[float]]:
+            """Evaluate the observables for the ansatz parameters ``theta``.
+
+            Args:
+                theta: The ansatz parameters.
+
+            Returns:
+                The observables evaluated at the ansatz parameters.
+            """
+            value_dict = dict(zip(ansatz_parameters, theta))
+            sampled = self._sampler.convert(converted, params=value_dict)
+            return sampled.eval()
+
+        return evaluate_observables
+
+    def evolve(self, evolution_problem: EvolutionProblem) -> EvolutionResult:
+        """
+        Args:
+            evolution_problem: The evolution problem containing the hamiltonian, total evolution
+                time and observables to evaluate.
+
+        Returns:
+            A result object containing the evolution information and evaluated observables.
+
+        Raises:
+            ValueError: If the evolution time is not positive or the timestep is too small.
+        """
+        time = evolution_problem.time
+        observables = evolution_problem.aux_operators
+        hamiltonian = evolution_problem.hamiltonian
+
+        if not 0 < self.timestep <= time:
+            raise ValueError(
+                f"The time step ({self.timestep}) must be larger than 0 and smaller equal "
+                f"the evolution time ({time})."
+            )
+
+        # get the function to evaluate the observables for a given set of ansatz parameters
+        evaluate_observables = self._get_observable_evaluator(observables)
```

Comment:
Can we somehow leverage the existing aux ops evaluator class used by other algorithms in the framework?

Reply:
Let's create an issue to improve `aux_ops_evaluator`.

Reply:
See https://github.com/Qiskit/qiskit-terra/issues/8347